torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,746 @@
+    and then indexes from this pyramid to create correlation features.
+    The "indexing" of a given centroid pixel x' is done by concatenating its surrounding row neighbours
+    within radius
+    """
+
+    def __init__(self, *, num_levels: int = 4, radius: int = 4):
+        super().__init__()
+        self.num_levels = num_levels
+        self.radius = radius
+
+        self.corr_pyramid: List[Tensor] = [torch.tensor(0)]  # useless, but torchscript is otherwise confused :')
+
+        self.out_channels = num_levels * (2 * radius + 1)
+
+    def build_pyramid(self, fmap1: Tensor, fmap2: Tensor):
+        """Build the correlation pyramid from two feature maps.
+
+        The correlation volume is first computed as the dot product of each pair (pixel_in_fmap1, pixel_in_fmap2) on the same row.
+        The last 2 dimensions of the correlation volume are then pooled num_levels times at different resolutions
+        to build the correlation pyramid.
+        """
+
+        if fmap1.shape != fmap2.shape:
+            raise ValueError(
+                f"Input feature maps should have the same shape, instead got {fmap1.shape} (fmap1.shape) != {fmap2.shape} (fmap2.shape)"
+            )
+        corr_volume = self._compute_corr_volume(fmap1, fmap2)
+
+        batch_size, h, w, num_channels, _ = corr_volume.shape
+        corr_volume = corr_volume.reshape(batch_size * h * w, num_channels, 1, w)
+        self.corr_pyramid = [corr_volume]
+        for _ in range(self.num_levels - 1):
+            corr_volume = F.avg_pool2d(corr_volume, kernel_size=(1, 2), stride=(1, 2))
+            self.corr_pyramid.append(corr_volume)
+
+    def index_pyramid(self, centroids_coords: Tensor) -> Tensor:
+        """Return correlation features by indexing from the pyramid."""
+        neighborhood_side_len = 2 * self.radius + 1  # see note in __init__ about out_channels
+        di = torch.linspace(-self.radius, self.radius, neighborhood_side_len)
+        di = di.view(1, 1, neighborhood_side_len, 1).to(centroids_coords.device)
+
+        batch_size, _, h, w = centroids_coords.shape  # _ = 2 but we only use the first one
+        # We only consider 1d and take the first dim only
+        centroids_coords = centroids_coords[:, :1].permute(0, 2, 3, 1).reshape(batch_size * h * w, 1, 1, 1)
+
+        indexed_pyramid = []
+        for corr_volume in self.corr_pyramid:
+            x0 = centroids_coords + di  # end shape is (batch_size * h * w, 1, side_len, 1)
+            y0 = torch.zeros_like(x0)
+            sampling_coords = torch.cat([x0, y0], dim=-1)
+            indexed_corr_volume = grid_sample(corr_volume, sampling_coords, align_corners=True, mode="bilinear").view(
+                batch_size, h, w, -1
+            )
+            indexed_pyramid.append(indexed_corr_volume)
+            centroids_coords = centroids_coords / 2
+
+        corr_features = torch.cat(indexed_pyramid, dim=-1).permute(0, 3, 1, 2).contiguous()
```

Comment:
Out of curiosity, why the `contiguous()` call? Is it for performance reasons? Have you measured it personally or found it on the original implementation?

Reply:
I follow the raft implementation for this: https://github.com/pytorch/vision/blob/main/torchvision/models/optical_flow/raft.py#L392

Reply:
I have tested removing the `.contiguous()` call, and seems like there is no error nor significant change in performance (there is only a very small change on the output result but it still within 1e-6).

Take note however that both the [torchvision's RAFT implementation](https://github.com/pytorch/vision/blob/main/torchvision/models/optical_flow/raft.py#L392) and the [original RAFT-Stereo implementation](https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/corr.py#L146) have this contiguous call.

Reply:
Let's leave and benchmark later.